### ISSUE TITLE
Continuous polling in Piston event loop, optimization

### ISSUE THREAD

**Issue by christolliday** [PistonDevelopers/conrod#802]
The core event loop used by Conrod seems to be a constant fixed rate loop where widgets are continuously polling for input events and continuously being updated, whether or not any events occur. This seems to be the cause of relatively high CPU usage even for a completely static GUI.

Am I misinterpreting the main loop, and if so, what accounts for the CPU usage of an idle Conrod app?

The best documentation for the Piston event loop I could find is here: https://github.com/PistonDevelopers/piston/issues/193

That type of event model makes sense for some games, where a continuously changing game environment is assumed, but not for a typical GUI app or even for many simpler games.

At any given point in time when there is no user input, no animations, no window resizing or any other events happening that would trigger UI changes, the CPU usage of a simple Conrod app should be zero or about to be zero. This would apply to all the current examples.

In addition to generally wasting resources, the current event model makes Conrod impractical for many use cases, especially for apps on battery constrained devices like smartphones or embedded devices.

At first glance, and at a very high level, I think moving to a more event driven design would involve:
- Breaking up the update method on widgets into two, one for receiving events and changing the state, and one for updating the widget. The update method would then only be called if an event took place which changed the widgets state.
- Replacing the continuous Piston event loop with something that only reacts to events, forwards them to widgets, and checks if updating and redrawing is necessary after any widget event callbacks are called
- For animations, allowing widgets to register for continuous, fixed rate event callbacks

Does this make sense and is it at all practical? Would such a change be consistent with Conrods immediate mode style?

**Comment by mitchmindtree** [comment]
Thanks for the issue, definitely keen to optimise the idle CPU usage however we can.

> The core event loop used by Conrod seems to be a constant fixed rate loop where widgets are continuously polling for input events and continuously being updated, whether or not any events occur.

Conrod itself doesn't require any specific core event loop. It interacts with an application's loop at three distinct points:

- `Ui::handle_event` can be thought of as the input, where events are received, interpreted as `event::Ui`s and pushed to a queue.

- `Ui::set_widgets` is the stage at which all widgets are instantiated (updated). During this call, the events that have been queued via calls to `Ui::handle_event` are made available to each widget and can be accessed and filtered by widgets in their `Widget::update` method. The events are then cleared at the end of this stage when the `UiCell` is `drop`ped.

  Conrod itself doesn't require that `Ui::set_widgets` is called at a fixed rate - for example, you can insert some condition where `Ui::set_widgets` is only called if some interaction event (mouse, resize, etc) has been passed to `Ui::handle_event`. I'd be happy to see the examples updated to do this and see if we can save some cycles. Perhaps we could even add an alternative to `Ui::set_widgets` like `Ui::set_widgets_if_interacted` that returns `Some(UiCell)` if some interaction event has been received? To support animations, we could perhaps add some `Widget` method to allow widgets to signal whether or not they're currently animating, similar to as you've mentioned?

- `Ui::draw` / `Ui::draw_if_changed` produces conrod's output - a sequence of `render::Primitive`s that can be drawn to screen however the user wishes. We provide a piston `draw` backend to make it easier to get started and hope to provide a glium-specific one once we can get the `glutin_glium.rs` example going (this should be easy and much more efficient than the generic piston one when finished, but I don't have a lot of GL experience and haven't had a chance to finish it off yet).

> The best documentation for the Piston event loop I could find

Conrod isn't specifically tied to the piston event loop - the main two reasons `piston_window` is used in most of the examples are:

- it is easy to set up compared to other windows and
- Conrod *used* to depend on piston-graphics and the piston event loop, but it no longer does. Today, it takes a it's own `event::Raw` type as input (via `Ui::handle_event`) and outputs the `render::Primitives` struct (via `Ui::draw`/`Ui::draw_if_changed`).

I'm keen to see us introduce more examples using different window-loop+graphics combinations (i.e. glutin+glium, glfw+gfx, etc) to demonstrate the current flexibility properly.

Hmmm your following suggestions seem like they might be based on this assumption:

> ... Conrod seems to be a constant fixed rate loop

which I think I've described that this isn't really the case - it just happens to be how the examples are implemented which, as you've pointed out, is definitely not as efficient as they could be! Am I interpreting your comment correctly? Let me know if I'm missing the point!

If we implement the changes that I've mentioned above and continue to run into issues with idle CPU in the examples we should use some profiling tools to see exactly where cycles are being spent so we know exactly what's going on and where we can improve.

**Comment by christolliday** [comment]
Ah ok great, that's actually a relief to hear about not being tied to the Piston event loop. I should have dug a little deeper since I was basing this just on the examples. It can be difficult to tell what is a necessary part of the architecture from what is just incidental to the examples. Thanks for clearing that up, I think it's good to have it documented here regardless.

I think it would be good to demonstrate using a non-piston event loop in the examples. I'm going to start looking into it and I could add an example if you think it's a good idea.

I'd even go so far as to say as they all should be moved off Piston, because what each example is demonstrating would be better implemented without a fixed rate loop. Then a new example could be added that is built on a Piston event loop that is more suited to that and that could potentially implement the event filtering you described.

I guess part of my confusion was from the fact that right now there are no examples of non-input driven animations in Conrod right now, correct? Right now if an animation not tied to user input was added, you would have to either use the Piston event loop or write up your own loop to re-instantiate the entire UI at a fixed rate for the duration of the animation, is that right?
As an example, a blinking cursor in a text box would be a lot more complicated to implement with a custom event loop than it probably should be, and would still waste some resources re-instantiating the entire UI whenever a cursor blink happens.

So in terms of general optimization, I have a few questions:
- If an event occurs that updates one widget, is it necessary for `Widget::update` to be called to re-instantiate every widget?
- A widgets `state` is used to detect changes and trigger redraw of the lower level graphics primitives, is that correct?
- Could it be possible that not just drawing but also instantiating the widget is triggered by changes to `state`? In this model, `state` would be changed by the event handling part of `Widget::update` being split off into it's own function.
- Is the current design partly based on the idea that `Widget::update` doesn't always trigger a redraw and is expected to be cheap compared to the drawing?

**Comment by christolliday** [comment]
To be clear, since I did conflate two separate issues, the first issue about the event loop I think just needs an example that demonstrates an alternative event loop, and maybe an example that demonstrates an animation in a non-continuous event loop.

The second issue is about optimizing what happens per update loop, which relates to the issue of providing some more retained mode benefits with an immediate mode API.

Sorry if I confused the issue by mixing the two up, I've done some more reading about the retained/immediate mode approach, and I think I've answered some of my own questions. I'll rephrase some of the per update optimization stuff in a separate question.

**Comment by mitchmindtree** [comment]
> I think it would be good to demonstrate using a non-piston event loop in the examples. I'm going to start looking into it and I could add an example if you think it's a good idea.

This would be great - I've been hoping to get around to this myself, but can only get so much done heh.

In order to do this, I think ideally we'd get the `glutin_glium.rs` example going first. By this I mean drawing the remaining un-handled render primitives to screen (`Rectangle`, `Lines`, `Polygon` and `Image`). Once this is done, we could move the drawing logic into the `backend/glium.rs` module, similar to how the piston drawing logic is in it's own backend module. This should make it easier to use glutin+glium in other examples too without having to duplicate too much code. The reasons why I'm thinking glutin+glium could be nice are

1. doesn't require a fixed loop
2. they're pure rust and wrap opengl so we don't require any external C libs or anything and we can remain portable and
3. drawing is much more efficient than with the piston graphics lib as we get direct access to vertex and index buffers, etc.

I noticed you mentioned you've already found answers to some of your following questions, but I'll try to answer just in case it's useful!

> If an event occurs that updates one widget, is it necessary for Widget::update to be called to re-instantiate every widget?

Yeah this is correct, though I think it's important to distinguish how widgets are instantiated in conrod as it's a little different to other GUIs.

The idea behind conrod's "immediate" API design approach is that the UI exists as a *result* or *reflection* of the data or state that exists in a user's application, rather than existing as a separate "retained" entity that the user needs to manually keep synchronised with their app's state. To do this, the idea is that instantiating widgets should be as cheap as possible, only ever taking `Copy` data or borrowed data as inputs. If dynamic allocations or expensive operations are necessary, their results should be cached within the widget's associated `State` type so that they may be re-used upon each update. Ideally, a widget being instantiated that hasn't changed or received any events since it was last instantiated should at most compare a small amount of stack data. In turn, conrod's widget instantiation *should* be very cheap. Even so, there's definitely still room for improvement and their are some areas where we might be able to be avoid some more work. Again though, doing some profiling would be our best bet to find these areas.

> A widgets state is used to detect changes and trigger redraw of the lower level graphics primitives, is that correct?

Yeah, if a widget's `Style` or `State` has changed between updates it triggers a redraw. Something I've been thinking about lately is that we might be able to save some cycles by only doing these checks for the "primitive" widgets, as they are the only ones that actually have a visual representation. I doubt we'd save much, though it might be worth experimenting if we get time.

> Could it be possible that not just drawing but also instantiating the widget is triggered by changes to state? In this model, state would be changed by the event handling part of Widget::update being split off into it's own function.

Hmm I'm not sure if splitting up the `update` and `event` handling would make a lot of sense for conrod's API, though now that you mention it we might be able to save some work by considering events more closely. I.e. when `Widget::set` is called, perhaps we could have some check that only calls `Widget::update` if either:

1. there is one or more pending events for either that widget or one of its children or
2. the widget doesn't yet exist within the widget graph

We might be able to save a fairly significant amount of work here, though I'd have to give it some more thought. One of the benefits of conrod's current design is that widgets automatically synchronise with changes to application data, even if there are no associated UI events. For example, the position of an overlay for a moving entity in a game, or an amplitude visualisation in an audio application. We'd have to provide a way for these widgets to continue to `update` regardless of pending events. Perhaps this could be a part of the hypothetical "animated widgets" API we touched on above?

> Is the current design partly based on the idea that Widget::update doesn't always trigger a redraw and is expected to be cheap compared to the drawing?

Yeah, I think I covered this in my reply to a previous question, but yes, `Widget::update`s should be as cheap as possible and cache using the associated `State` type where possible.

**Comment by christolliday** [comment]
Yep this absolutely helps a lot, thanks for the info.

It sounds like there's a lot of avenues for optimization, in terms of prioritization my biggest concern right now would be battery life since I'm interested in potentially using Conrod on mobile devices. In that use case (and in general) there's a big difference between waking up the CPU 60 times a second vs only when something changes. For this use case, animations are the only thing that would cause UI code to run continously but there's not much you can do about it besides trying to make animations only run intermittently.

For optimizations other than that, I agree there should be some profiling done to find the biggest costs.

I think there will eventually have to be some kind of API for animated widgets. Fitting it into the current API with minimal changes could be as simple as a widget setting a flag during update. Then it would continue being updated at a fixed rate as long as the flag is set.

So right now I'm trying this with Piston just because getting glium+glutin up to feature parity with Piston as a backend sounds like a bigger task, but it does sound like probably a better path long term.

I started looking into improving the Piston event loop, using an custom event loop that uses the blocking `GlutinWindow::wait_events` rather than `GlutinWindow::poll_events` interleaved with calls to `sleep`. It seemed like the quickest path to have a non polling main event loop, that maybe could be adapted to another backend.

Right now I have a basically hacked together fork of Piston that waits for events so that it will actually use 0 CPU while running in the background or when nothing is happening. I just need to figure out how to integrate animations with it then figure out how to upstream it.
It seems like if I can make animations run along side a blocking event loop, then just having animations enabled all the time will cause the same behaviour and it could be merged into mainline Piston? I'll see how it turns out then I could open an issue on the Piston repo.

**Comment by mitchmindtree** [comment]
Closing this following #909.
